src/diffusers/hooks/teacache.py

```diff
@@ -0,0 +1,1021 @@
+        state_manager (StateManager):
+            Manages TeaCacheState across forward passes, maintaining counters and cached values.
+    """
+
+    _is_stateful = True
+
+    def __init__(self, config: TeaCacheConfig):
+        super().__init__()
+        self.config = config
+        # Set default rescale_func with config coefficients (will be updated in initialize_hook if needed)
+        # This ensures rescale_func is always valid, even if initialize_hook isn't called (e.g., in tests)
+        default_coeffs = config.coefficients if config.coefficients else _MODEL_COEFFICIENTS["Flux"]
+        self.coefficients = default_coeffs
+        self.rescale_func = self._create_rescale_func(default_coeffs)
+        self.state_manager = StateManager(TeaCacheState, (), {})
+        self.extractor_fn = None
+        self.model_type = None
+
+    @staticmethod
+    def _create_rescale_func(coefficients):
+        """Create polynomial rescale function from coefficients.
+        
+        Evaluates: c[0]*x^4 + c[1]*x^3 + c[2]*x^2 + c[3]*x + c[4]
+        """
+        def rescale(x):
+            return coefficients[0] * x**4 + coefficients[1] * x**3 + coefficients[2] * x**2 + coefficients[3] * x + coefficients[4]
+        return rescale
+
+    def _maybe_reset_state_for_new_inference(self, state, module, reset_encoder_residual=False):
+        """Reset state if we've completed all steps (start of new inference run).
+        
+        Also initializes num_steps on first timestep if not set.
+        
+        Args:
+            state: TeaCacheState instance.
+            module: The transformer module.
+            reset_encoder_residual: If True, also reset previous_residual_encoder (for CogVideoX).
+        """
+        # Reset counter if we've completed all steps (new inference run)
+        if state.cnt == state.num_steps and state.num_steps > 0:
+            logger.info("TeaCache inference completed")
+            state.cnt = 0
+            state.accumulated_rel_l1_distance = 0.0
+            state.previous_modulated_input = None
+            state.previous_residual = None
+            if reset_encoder_residual:
+                state.previous_residual_encoder = None
+
+        # Set num_steps on first timestep if not already set
+        if state.cnt == 0 and state.num_steps == 0:
+            # Priority: config value > callback > module attribute
+            if self.config.num_inference_steps is not None:
+                state.num_steps = self.config.num_inference_steps
+            elif self.config.num_inference_steps_callback is not None:
+                state.num_steps = self.config.num_inference_steps_callback()
+            elif hasattr(module, "num_steps"):
+                state.num_steps = module.num_steps
+            
+            if state.num_steps > 0:
+                logger.info(f"TeaCache: Using {state.num_steps} inference steps")
+
+    def initialize_hook(self, module):
+        self.state_manager.set_context("teacache")
```

Comment:
Cache context is typically set in the denoising loop? I think in this case, both conditional and unconditional branches would write to the same cache state when using CFG.